ss/pebbledb/db.go

```diff
@@ -186,13 +187,90 @@ func (db *Database) ApplyChangeset(version int64, cs *proto.NamedChangeSet) erro
 	return b.Write()
 }
 
-// Prune for the PebbleDB SS backend is currently not supported. It seems the only
-// reliable way to prune is to iterate over the desired domain and either manually
-// tombstone or delete. Either way, the operation would be timely.
-//
-// See: https://github.com/cockroachdb/cockroach/blob/33623e3ee420174a4fd3226d1284b03f0e3caaac/pkg/storage/mvcc.go#L3182
+// Prune attempts to prune all versions up to and including the current version
+// Get the range of keys, manually iterate over them and delete them
 func (db *Database) Prune(version int64) error {
-	panic("not implemented!")
+	itr, err := db.storage.NewIter(nil)
```

Comment:
Shouldn't the iterator be `db.NewIterator(nil, nil, version)`? This way you only iterate over keys <= version and you can avoid having to compare versions in the loop.

Reply:
You have to iterate to version higher than prune version to determine if a key can be deleted or not. We can't blindly just delete the key if the key version is smaller than prune height

Reply:
> We can't blindly just delete the key if the key version is smaller than prune height

Why not? This is exactly what pruning is supposed to do and what all SS backends do. Delete absolutely everything <= PruneVersion

Reply:
Nope, we discussed last week and notice this is wrong. Think about it, if you have any data that only has a single version and never got updated, you will end up pruning all of them which you shouldn't. Because these keys are not deleted at all, so you need to keep a latest version for each key, that's what RocksDB is doing

Reply:
For example, let's say we did the following operations:
- Set k1, v1 at height 100
- Set k1, v2 at height 200

This actually means:
- From height 1 - 99, k1 doesn't exist
- From height 100 - 199, k1 has value v1
- From height 200 -> latest k1 has value v2

If the pruning policy set keep-recent=100, then
- At chain height 101 we do nothing
- At chain height 250, we still do nothing, because keep recent 100 means (150,250], and k1 has v1 exist until height 199, so we can not delete k1 yet. 
- At chain height 300, we can delete k1:v1, because we only keep block 200-300 now, which k1 has value of v2
- At chain height 400, we do nothing, because k1 is not deleted, if something query k1 at height 350, should still be querable

Reply:
I see. Yes, this makes sense from query perspective. But I'm not 100% certain RocksDB `full_history_ts_low` behaves this way. Have you confirmed it does? If so, I think we'll adjust the sqlite implementation as well.